### PR TITLE
[Merged by Bors] - feat: determinant of left multiplication

### PR DESCRIPTION
**Pull request by YaelDillies**
From FLT

Co-authored-by: Kevin Buzzard <k.buzzard@imperial.ac.uk>

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/LinearAlgebra/Determinant.lean [hunk 1]
```diff
@@ -324,6 +324,11 @@ theorem bot_lt_ker_of_det_eq_zero {𝕜 : Type*} [Field 𝕜] [Module 𝕜 M] {f
   simp only [bot_lt_iff_ne_bot, Classical.not_not, ← isUnit_iff_ker_eq_bot] at hf
   exact isUnit_iff_ne_zero.1 (f.isUnit_det hf)
 
+@[simp] lemma det_mul (a : R) : (mul R R a).det = a := by
```
[the remark this answers is not in the document]
Comment (alreadydone) [reply Mathlib/LinearAlgebra/Determinant.lean:327]: [LinearMap.det_smul](https://leanprover-community.github.io/mathlib4_docs/Mathlib/LinearAlgebra/Determinant.html#LinearMap.det_smul) assumes Field while it should really assume M is finite free ...

[the remark this answers is not in the document]
Comment (b-mehta) [reply Mathlib/LinearAlgebra/Determinant.lean:327]: That lemma also suggests that this one is named inconsistently

[the remark this answers is not in the document]
Comment (eric-wieser) [reply Mathlib/LinearAlgebra/Determinant.lean:327]: > [LinearMap.det_smul](https://leanprover-community.github.io/mathlib4_docs/Mathlib/LinearAlgebra/Determinant.html#LinearMap.det_smul) assumes Field while it should really assume M is finite free ...

#19885

[the remark this answers is not in the document]
Comment (eric-wieser) [reply Mathlib/LinearAlgebra/Determinant.lean:327]: This lemma should probably be called `det_mulLeft` and stated using `mulLeft`.

File: Mathlib/LinearAlgebra/Determinant.lean [hunk 2]
```diff
@@ -329,6 +329,11 @@ theorem bot_lt_ker_of_det_eq_zero {𝕜 : Type*} [Field 𝕜] [Module 𝕜 M] {f
   simp only [bot_lt_iff_ne_bot, Classical.not_not, ← isUnit_iff_ker_eq_bot] at hf
   exact isUnit_iff_ne_zero.1 (f.isUnit_det hf)
 
+@[simp] lemma det_mulLeft (a : R) : (mulLeft R a).det = a := by
```
Comment (eric-wieser) [inline Mathlib/LinearAlgebra/Determinant.lean:332]: Can you add the `mulRight` version too?

**Comment by YaelDillies** [comment]
I thought the resolution was your suggestion of stating the result with `mulLeft`. Given that `det_smul` already exists, I don't really see the point of "generalising" (it's not syntactically a generalisation) my result to scalar multiplication.

**Comment by eric-wieser** [comment]
`LinearMap.det_smul` is about the result of scaling a linear map, not the result of taking the determinant of the map that corresponds to scaling (`DistribMulAction.toLinearMap`).

**Comment by YaelDillies** [comment]
> `LinearMap.det_smul` is about the result of scaling a linear map, not the result of taking the determinant of the map that corresponds to scaling

Yes, but `(a • ·) = a • id`, right? So you can get one from the other. That was my reasoning

**Comment by b-mehta** [comment]
No, the point of my suggestion is that your lemma is about multiplication by an element of R as a map R -> R, whereas I'd like to generalise to multiplication by an element of R as a map M -> M

**Comment by YaelDillies** [comment]
Yes but that's not a syntactic generalisation, so it would have to be a separate lemma

**Comment by eric-wieser** [comment]
It would be a separate lemma from which your lemma can be proven directly.

**Comment by YaelDillies** [comment]
Okay, then which of the 4-5 spellings of this linear map should I be writing it for?

**Comment by eric-wieser** [comment]
`DistribMulAction.toLinearMap` is the most general one

**Comment by YaelDillies** [comment]
Okay, and what should it state? Scalar multiplication by a real number `r` has determinant `r` over the reals, but `r ^ 2` over the complex numbers.

**Comment by eric-wieser** [comment]
How about we add
```
/-- When the function is over the base ring, the determinant is the evaluation at 1. -/
@[simp] lemma det_ring (f : R →ₗ[R] R) : f.det = f 1 := by
  classical
  rw [← det_toMatrix ⟨(Finsupp.LinearEquiv.finsuppUnique R R Unit).symm⟩, Matrix.det_unique]
  rfl
```
instead, from which your lemmas follow with `by simp`?

**Comment by YaelDillies** [comment]
That's a great idea, although I don't really like the name `det_ring`

File: Mathlib/LinearAlgebra/Determinant.lean [hunk 3]
```diff
@@ -329,6 +329,15 @@ theorem bot_lt_ker_of_det_eq_zero {𝕜 : Type*} [Field 𝕜] [Module 𝕜 M] {f
   simp only [bot_lt_iff_ne_bot, Classical.not_not, ← isUnit_iff_ker_eq_bot] at hf
   exact isUnit_iff_ne_zero.1 (f.isUnit_det hf)
 
+/-- When the function is over the base ring, the determinant is the evaluation at `1`. -/
+@[simp] lemma det_ring (f : R →ₗ[R] R) : f.det = f 1 := by
+  classical
+  rw [← det_toMatrix ⟨(Finsupp.LinearEquiv.finsuppUnique R R Unit).symm⟩, Matrix.det_unique]
```
Comment (eric-wieser) [inline Mathlib/LinearAlgebra/Determinant.lean:335]: Any idea why `(Basis.singleton Unit R)` doesn't work here? Are we missing a simp lemma? Are they defeq?

**Comment by eric-wieser** [comment]
> I don't really like the name `det_ring`

I chose it given we have `LinearMap.ext_ring`

**Review by jcommelin — COMMENTED** [review]
I'm happy with this as is, but also fine with the alternative proof that @YaelDillies provided.

The name `det_ring` is maybe not perfect, but I don't have a better suggestion.

So I'm going to

bors d+

**Comment by YaelDillies** [comment]
bors merge

**Comment by eric-wieser** [comment]
bors r-

bors d+

File: Mathlib/LinearAlgebra/Matrix/ToLin.lean [hunk 4]
```diff
@@ -561,6 +561,11 @@ theorem LinearMap.toMatrix_id : LinearMap.toMatrix v₁ v₁ id = 1 := by
 theorem LinearMap.toMatrix_one : LinearMap.toMatrix v₁ v₁ 1 = 1 :=
   LinearMap.toMatrix_id v₁
 
+@[simp]
+lemma LinearMap.toMatrix_singleton {ι : Type*} [Unique ι] (f) (i j : ι) :
```
[the remark this answers is not in the document]
Comment (YaelDillies) [reply Mathlib/LinearAlgebra/Matrix/ToLin.lean:565]: Oops yes I thought I had

**Comment by YaelDillies** [comment]
bors merge

**Pull request closed without merging**